OSDOCS-19969: Gateway API backendTLSPolicy for securing HTTPRoutes (#112677)
openshift/openshift-docs: modules/about-backend-tls-validation.adoc

// Module included in the following assemblies:
//
// * networking/ingress_load_balancing/configuring_gateway_api/securing-httproutes.adoc

:_mod-docs-content-type: CONCEPT
[id="about-backend-tls-validation_{context}"]
= About backend TLS validation

[role="_abstract"]

> To ensure secure end-to-end encryption, you can use a `BackendTLSPolicy` to validate your backend certificates. This validation ensures your gateway securely connects to the correct destination pod.

Reviewer: ```suggestion
To ensure secure end-to-end encryption, you can use a `BackendTLSPolicy` custom resource (CR) to validate your backend certificates. This validation ensures that your gateway securely connects to the correct destination pod.
```
mixed use of generic "gateway" and referring to the Gateway object could be confusing

Author: Thanks

To use re-encrypt termination in Gateway API, you must combine listener edge termination with a `BackendTLSPolicy` attached to the service.

> [NOTE]

Reviewer: ```suggestion
[IMPORTANT]
```

Author: done

====

> Unlike standard {product-title} routes, re-encrypt for `HTTPRoutes` requires a listener configured with edge termination on the gateway and a `BackendTLSPolicy` that targets the service.

Reviewer: ```suggestion
Re-encrypt for `HTTPRoutes` requires a listener configured with edge termination on the gateway and a `BackendTLSPolicy` CR that targets the service.
```
"unlike..." just adds words; it does not enhance the meaning and buries the lede
i would remove this preamble throughout (you might also consider making this a snippet that you can include in all of the places where you want it)

Author: Thanks!

Author: Okay, so I was looking into adding the snippet and I think I'm going to hold off for now. Since the important block only appears twice across the docs and our core docs checklist requires managerial approval to add new attribute files, I think for now it would be best to just manually update the text in both locations.

====

> `BackendTLSPolicy` describes one or more (up to 16) targets of the policy (`targetRefs`) and how the backend certificate should be validated (`validation`).

Reviewer: ```suggestion
A `BackendTLSPolicy` CR describes up to 16 targets of the policy described in the `targetRefs` parameter. The policy also determines how the backend certificate should be validated.
```
no parens (IBM style)

Author: Yikes! Sorry, should know better

> For validation, `BackendTLSPolicy` specifies two options to validate one or more backend certificates for a TLS connection. You can configure only one of the following options per policy:

Reviewer: ```suggestion
The `BackendTLSPolicy` CR specifies two options to validate backend certificates for a TLS connection. You must configure only one of the following options per policy:
```
don't need to repeat "one or more"
if you want TLS, this would be a "must"

Author: done, thanks

`caCertificateRefs`:: 

> Use this option for explicit CA certificates. These refer to one or more (up to 8) PEM-encoded TLS certificates configured within a `configMap` object. The `configMap` object must reside in the same namespace as the `BackendTLSPolicy`. Cross-namespace references via a `ReferenceGrant` are not supported.

Reviewer: ```suggestion
Use this option for explicit certificate authority (CA) certificates. The `caCertificateRefs` parameter can refer to up to 8 PEM-encoded TLS certificates configured within a config map. The config map must reside in the same namespace as the `BackendTLSPolicy`. Cross-namespace references with a `ReferenceGrant` are not supported.
```
do not use via (see IBM style guide)
config map or `ConfigMap` object see https://redhat-documentation.github.io/supplementary-style-guide/#config-map

Author: thanks

`wellKnownCACertificates`:: 
Use this option for trusted system certificates.

> Validation also specifies a `hostname` (Server Name Indication (SNI)) that must match the certificate served by the destination pod, unless `subjectAltNames` (Subject Alternative Name (SAN)) is also specified. You can specify up to five `subjectAltNames`. 

Reviewer: ```suggestion
Validation also specifies a Server Name Indication (SNI) with the `hostname` parameter. The SNI must match the certificate served by the destination pod, unless a Subject Alternative Name (SAN) is also specified. You can specify up to five SANs in the `subjectAltNames` parameter. 
```
no parens...

> The choice of whether to use `hostname` or `subjectAltNames` depends on the certificate served by the targeted backend pods. If the pod’s certificate uses SANs, the `BackendTLSPolicy` must have at least one matching `subjectAltName` to successfully connect using TLS. The `subjectAltNames` field must include a `type`, either `hostname` or Uniform Resource Identifier (`URI`), and can be specified as domain names, URIs, or a combination of both.

Reviewer: ```suggestion
The choice of whether to use `hostname` or `subjectAltNames` depends on the certificate served by the targeted backend pods. If the pod certificate uses SANs, the `BackendTLSPolicy` must have at least one matching `subjectAltName` to successfully connect using TLS. The `subjectAltNames` field must include a `type`, either `hostname` or Uniform Resource Identifier (`URI`), and can be specified as domain names, URIs, or a combination of both.
```
no possessives (IBM)

> The choice of whether to use `hostname` or `subjectAltNames` depends on the certificate served by the targeted backend pods. If the pod’s certificate uses SANs, the `BackendTLSPolicy` must have at least one matching `subjectAltName` to successfully connect using TLS. The `subjectAltNames` field must include a `type`, either `hostname` or Uniform Resource Identifier (`URI`), and can be specified as domain names, URIs, or a combination of both.

Reviewer: ```suggestion
The choice of whether to use SNIs  or SANs depends on the certificate served by the targeted backend pods. If the pod certificate uses SANs, the `BackendTLSPolicy` CR must have at least one matching `subjectAltName` parameter value to successfully connect using TLS. The `subjectAltNames` field must include a `type`. The type can be either a `hostname` or Uniform Resource Identifier (`URI`), and can be specified as a domain name, URI, or a combination of both.
```

> The choice of whether to use `hostname` or `subjectAltNames` depends on the certificate served by the targeted backend pods. If the pod’s certificate uses SANs, the `BackendTLSPolicy` must have at least one matching `subjectAltName` to successfully connect using TLS. The `subjectAltNames` field must include a type, either `hostname` or Uniform Resource Identifier (`URI`), and can be specified as domain names, URIs, or a combination of both.

Reviewer 2: `type` is also a variable name that must be set.
```suggestion
The choice of whether to use `hostname` or `subjectAltNames` depends on the certificate served by the targeted backend pods. If the pod’s certificate uses SANs, the `BackendTLSPolicy` must have at least one matching `subjectAltName` to successfully connect using TLS. The `subjectAltNames` field must include a `type`, either `hostname` or Uniform Resource Identifier (`URI`), and can be specified as domain names, URIs, or a combination of both.
```

Author: good catch! fixed

> [NOTE]

Reviewer: ```suggestion
[IMPORTANT]
```
because if you don't do this, the process fails

====

> Gateway API does not automatically evaluate the `hostname` as a subject alternative name (SAN). If your certificate requires the hostname to be validated as a SAN, you must explicitly add it to the `subjectAltNames` list.

Reviewer: ```suggestion
Gateway API does not automatically evaluate the `hostname` as a SAN. If your certificate requires the hostname to be validated as a SAN, you must explicitly add it to the `subjectAltNames` list.
```

Author: good point, fixed
